src/python/pants/backend/python/util_rules/pex_cli.py

```diff
@@ -130,6 +204,114 @@ async def download_pex_pex(pex_cli: PexCli, platform: Platform) -> PexPEX:
+
+    keyring_plugin_request_types = union_membership.get(PexKeyringConfigurationRequest)
+    if not keyring_plugin_request_types:
+        return _KeyringState(keyring_data_path=None)
+
+    credentials_responses: dict[str, list[tuple[str, tuple[str, str]]]] = {}
+    for keyring_plugin_request_type in keyring_plugin_request_types:
+        keyring_plugin_request = keyring_plugin_request_type()
+        keyring_plugin_response = await Get(  # noqa: PNT30: Only one provider expected.
+            PexKeyringConfigurationResponse, PexKeyringConfigurationRequest, keyring_plugin_request
+        )
+
+        keyring_plugin_credentials = keyring_plugin_response.credentials
+        if not keyring_plugin_credentials:
+            continue
+
+        for site, user_password in keyring_plugin_credentials.items():
+            if site not in credentials_responses:
+                credentials_responses[site] = []
+            credentials_responses[site].append((keyring_plugin_request_type.name, user_password))
+
+    # Short circuit if the providers did not supply any credentials.
+    if not credentials_responses:
+        return _KeyringState(keyring_data_path=None)
+
+    # Check for multiple responses for a single site. Keyring and our simulation of it only supports
+    # a single user/password per site.
+    credentials: list[tuple[str, str, str]] = []
+    for site, user_password_creds_by_provider in credentials_responses.items():
+        if len(user_password_creds_by_provider) > 1:
+            providers = [x[0] for x in user_password_creds_by_provider]
+            raise Exception(
+                f"Multiple keyring plugins returned responses for `{site}`. "
+                "The keyring support only supports a single user/password per site. "
+                f"The keyring plugins in conflict are: {', '.join(providers)}."
+            )
+        user_password_creds = user_password_creds_by_provider[0][1]
+        credentials.append((site, user_password_creds[0], user_password_creds[1]))
+    credentials.sort()
+
+    # Write the credentials to a file based on the current run ID.
+    content = StringIO()
+    content.write("declare -A pants_keyring_credentials\n")
+    for site, user, password in credentials:
+        content.write(
+            f"pants_keyring_credentials[{_always_shlex_quote(site)}]={_always_shlex_quote(user)}\n"
+        )
+        site_user = f"{site}:{user}"
+        content.write(
+            f"pants_keyring_credentials[{_always_shlex_quote(site_user)}]={_always_shlex_quote(password)}\n"
+        )
+
+    # TODO: Consider removing the run ID so that this value does not contribute to the cache key
+    # and cause unneeded invalidation of build actions as it changes for each run.
+    keyring_data_path = build_root.pathlib_path / ".pants.d" / "keyring" / str(run_id) / "data.sh"
+    keyring_data_path.parent.mkdir(parents=True, exist_ok=True)
+    keyring_data_path.write_bytes(content.getvalue().encode())
```

Comment:
Should we not ensure that this file has no greater than 0600 perms?

Reply:
And should we not delete it when the process returns?

Reply:
See my comment at https://github.com/pantsbuild/pants/pull/21852#discussion_r1921479357 about concurrent runs. Would love your input on that question.